File: tests/scala/src/main/scala/sertests/MemoryFormatReader.scala
package sertests

import zio.{IO, UIO, Chunk}
import dev.argon.verilization.scala_runtime.FormatReader
import java.io.{IOException, EOFException}
import java.nio.{BufferUnderflowException, ByteBuffer, ByteOrder}

final class MemoryFormatReader(data: ByteBuffer) extends FormatReader[Any, EOFException] {

    private def catchErrors(ex: Throwable): IO[EOFException, Nothing] =
        ex match {
            case ex: BufferUnderflowException => IO.fail(new EOFException())
            case _ => IO.die(ex)
        }

    override def readByte(): IO[EOFException, Byte] =
        IO.effect { data.get() }.catchAll(catchErrors)

    override def readShort(): IO[EOFException, Short] =
        IO.effect { data.getShort() }.catchAll(catchErrors)

    override def readInt(): IO[EOFException, Int] =
        IO.effect { data.getInt() }.catchAll(catchErrors)

    override def readLong(): IO[EOFException, Long] =
        IO.effect { data.getLong() }.catchAll(catchErrors)

    override def readBytes(count: Int): IO[EOFException, Chunk[Byte]] =
        IO.effect {
            val buffer = new Array[Byte](count)
            data.get(buffer)
            Chunk.fromArray(buffer)
        }.catchAll(catchErrors)

    def isEOF: UIO[Boolean] =
        IO.effectTotal { !data.hasRemaining() }
}

File: tests/java/src/main/java/sertests/MemoryFormatReader.java
package sertests;

import dev.argon.verilization.runtime.FormatReader;
import java.io.IOException;
import java.io.EOFException;
import java.nio.BufferUnderflowException;
import java.nio.ByteBuffer;
import java.nio.ByteOrder;

public final class MemoryFormatReader implements FormatReader {

    public MemoryFormatReader(ByteBuffer data) {
        data.order(ByteOrder.LITTLE_ENDIAN);
        this.data = data;
    }

    private final ByteBuffer data;

    @Override
    public byte readByte() throws IOException {
        try {
            return data.get();
        }
        catch(BufferUnderflowException ex) {
            throw new EOFException();
        }
    }

	@Override
    public short readShort() throws IOException {
        try {
            return data.getShort();
        }
        catch(BufferUnderflowException ex) {
            throw new EOFException();
        }
    }

    @Override
	public int readInt() throws IOException {
        try {
            return data.getInt();
        }
        catch(BufferUnderflowException ex) {
            throw new EOFException();
        }
    }

    @Override
	public long readLong() throws IOException {
        try {
            return data.getLong();
        }
        catch(BufferUnderflowException ex) {
            throw new EOFException();
        }
    }

    @Override
	public byte[] readBytes(int count) throws IOException {
        byte[] buffer = new byte[count];
        data.get(buffer);
        return buffer;
    }

    public boolean isEOF() {
        return !data.hasRemaining();
    }
}
